sql/src/main/java/io/crate/operation/projectors/FetchProjector.java

```diff
@@ -230,18 +231,22 @@ public void fail(Throwable throwable) {
     private Row partitionedByRow(String index) {
         if (!partitionedBy.isEmpty() && PartitionName.isPartition(index)) {
             Object[] partitionValues;
-            synchronized (partitionValuesCacheLock) {
-                partitionValues = partitionValuesCache.get(index);
-                if (partitionValues == null) {
-                    List<BytesRef> partitionRawValues = PartitionName.fromStringSafe(index).values();
-                    partitionValues = new Object[partitionRawValues.size()];
-                    for (int i = 0; i < partitionRawValues.size(); i++) {
-                        partitionValues[i] = partitionedBy.get(i).type().value(partitionRawValues.get(i));
-                    }
-                    partitionValuesCache.put(index, partitionValues);
+            synchronized (partitionRowsCacheLock) {
+                Row partitionValuesRow = partitionRowsCache.get(index);
+                if (partitionValuesRow != null) {
+                    return partitionValuesRow;
                 }
             }
-            return new RowN(partitionValues);
+            List<BytesRef> partitionRowValues = PartitionName.fromStringSafe(index).values();
+            partitionValues = new Object[partitionRowValues.size()];
+            for (int i = 0; i < partitionRowValues.size(); i++) {
+                partitionValues[i] = partitionedBy.get(i).type().value(partitionRowValues.get(i));
+            }
+            Row partitionValuesRow = new RowN(partitionValues);
+            synchronized (partitionRowsCacheLock) {
+                partitionRowsCache.put(index, partitionValuesRow);
+            }
+            return partitionValuesRow;
         }
         return null;
```

Comment:
if index is not partitioned, an NULL entry must be put to the index->row map to avoid further lookups.
so map lookup should be done before checking if index is partitioned